src/site/content/en/reliable/offline-cookbook/index.md

```diff
@@ -0,0 +1,761 @@
+[update the cache entry](#on-network-response).
+
+However, this method has flaws. If the user has an intermittent or slow connection they'll have to
+wait for the network to fail before they get the perfectly acceptable content already on their
+device. This can take an extremely long time and is a frustrating user experience. See the next
+pattern, [Cache then network](#cache-then-network), for a better solution.
+
+```js
+self.addEventListener('fetch', function (event) {
+  event.respondWith(
+    fetch(event.request).catch(function () {
+      return caches.match(event.request);
+    }),
+  );
+});
+```
+
+### Cache then network {: #cache-then-network }
+
+<figure class="w-figure">
+  <img class="w-screenshot" alt="" src="ss-cache-then-network.png">
+</figure>
+
+**Ideal for:** Content that updates frequently. E.g. articles, social media timelines, game
+leaderboards.
+
+This requires the page to make two requests, one to the cache, one to the network. The idea is to
+show the cached data first, then update the page when/if the network data arrives.
+
+Sometimes you can just replace the current data when new data arrives (e.g. game leaderboard), but
+that can be disruptive with larger pieces of content. Basically, don't "disappear" something the
+user may be reading or interacting with.
+
+Twitter adds the new content above the old content and adjusts the scroll position so the user is
+uninterrupted. This is possible because Twitter mostly retains a mostly-linear order to content. I
+copied this pattern for [trained-to-thrill][ttt] to get content on screen as fast as possible, while
+displaying up-to-date content as soon as it arrives.
+
+**Code in the page:**
+
+```js
+var networkDataReceived = false;
+
+startSpinner();
+
+// fetch fresh data
+var networkUpdate = fetch('/data.json')
+  .then(function (response) {
+    return response.json();
+  })
+  .then(function (data) {
+    networkDataReceived = true;
+    updatePage(data);
+  });
+
+// fetch cached data
+caches
+  .match('/data.json')
+  .then(function (response) {
+    if (!response) throw Error('No data');
+    return response.json();
+  })
+  .then(function (data) {
+    // don't overwrite newer network data
+    if (!networkDataReceived) {
+      updatePage(data);
+    }
+  })
+  .catch(function () {
+    // we didn't get cached data, the network is our last hope:
+    return networkUpdate;
+  })
+  .catch(showErrorMessage)
+  .then(stopSpinner);
+```
+
+**Code in the Service Worker:**
+
+We always go to the network and update a cache as we go.
+
+```js
+self.addEventListener('fetch', function (event) {
+  event.respondWith(
+    caches.open('mysite-dynamic').then(function (cache) {
+      return fetch(event.request).then(function (response) {
+        cache.put(event.request, response.clone());
+        return response;
+      });
+    }),
+  );
+});
+```
+
+In [trained-to-thrill][ttt] I worked around this by using
```

Comment:
We need to update this somehow. XHR is about to go away. Even if Jake doesn't have time to take it out of Trained to Thrill, we need to call this out in some way.

"Back when I created trained to thrill, I worked around this by...

But XHR is going away. A contemporary solution using fetch would be..."

Reply:
This is not correct. XHR is here to stay. It may not be the primary choice today, but the API is by no means deprecated or anything.

Reply:
https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest